### PR TITLE
Qt: Fixing restore from system tray behaviour of main window

### PR DESCRIPTION
**Pull request by vajdaz**
When restoring the main window from the system tray, it always was in minimized state (at least on Windows). This patch should fix this. The window should restore to the same state as it was  at the moment of minimizing it into the tray.

May be related to issue #8225 

To reproduce the bug do following under Windows:
* In Options->Window enable "Minimize to the tray instead of the taskbar"
* Optionally enable "Minimize on close", too
* Minimize the main window
* Klick on the tray icon (or right click on the tray icon and select Show/Hide)

The main window's taskbar icon appears, but the window itself is not restored to its original size.

Expected behaviour: the window should appear.

### CODE REVIEW

File: src/qt/bitcoingui.cpp [hunk 1]
```diff
@@ -946,12 +945,25 @@ void BitcoinGUI::changeEvent(QEvent *e)
             QWindowStateChangeEvent *wsevt = static_cast<QWindowStateChangeEvent*>(e);
             if(!(wsevt->oldState() & Qt::WindowMinimized) && isMinimized())
             {
+                if (wsevt->oldState() & Qt::WindowMaximized)
+                {
+                    QTimer::singleShot(0, this, SLOT(showMaximized()));
+                }
+                else if (wsevt->oldState() & Qt::WindowFullScreen)
```
Comment (promag) [inline src/qt/bitcoingui.cpp]: Remove fullscreen?

File: src/qt/bitcoingui.cpp [hunk 2]
```diff
@@ -946,12 +945,25 @@ void BitcoinGUI::changeEvent(QEvent *e)
...
+                {
+                    QTimer::singleShot(0, this, SLOT(showNormal()));
+                }
                 QTimer::singleShot(0, this, SLOT(hide()));
                 e->ignore();
             }
         }
     }
 #endif
+    QMainWindow::changeEvent(e);
```
Comment (promag) [inline src/qt/bitcoingui.cpp]: Why?

**Review by promag — COMMENTED** [review]
This must be tested in linux too. BTW see #941.

File: src/qt/bitcoingui.cpp (hunk 1 above)
Comment (vajdaz) [reply src/qt/bitcoingui.cpp]: Removed.

File: src/qt/bitcoingui.cpp (hunk 2 above)
Comment (vajdaz) [reply src/qt/bitcoingui.cpp]: I thought, it makes senses to reorder. But I see, it makes no sense. Fixed.

**Comment by vajdaz** [comment]
I also have testetd on Kubuntu  (Plasma) and Ubuntu (Gnome), both 16.04. and Windows 10 Seems to be fine to me. However, on Gnome there is no system tray. So if you minimize to tray, you loose visual reference to your application. But this is nothing that this patch introduces, it is a conceptual problem that we have already.

File: src/qt/bitcoingui.cpp (hunk 2 above)
Comment (promag) [reply src/qt/bitcoingui.cpp]: You could have some reason, otherwise it's unrelated to your goal.

**Comment by laanwj** [comment]
I've restarted the travis build.

**Comment by promag** [comment]
@laanwj me too, about an hour ago. Was it still failing?

File: src/qt/bitcoingui.cpp (hunk 2 above)
Comment (vajdaz) [reply src/qt/bitcoingui.cpp]: The idea was, that theoretically we don't know what QMainWindow::changeEvent does. So if I want to add something "on top" of the default behaviour, I have to do it after that and not before. But this is kind of a academical thought. To be honest, I am not a Qt expert.

**Comment by laanwj** [comment]
> This must be tested in linux too. BTW see #941.

Yes we've had so much problems with this feature in the past...

**Comment by promag** [comment]
> it'd be better to remove it.

Only linux, or all?

**Comment by vajdaz** [comment]
Meanwhile I also read a about some past tray icon issues with different OSes. Seems that Qt can't do it right everywhere.

Anyway, I have the impression, that minimizing to tray is a common thing to do on Windows. However, on Linux this is not the case (see GNOME). On Mac I don't know.

What about keeping the minimize to tray feature only on Windows?

**Comment by laanwj** [comment]
> On Mac I don't know.

It's already disabled on MacOSX, along with the other window options:
```c++
#ifdef Q_OS_MAC
    /* remove Window tab on Mac */
    ui->tabWidget->removeTab(ui->tabWidget->indexOf(ui->tabWindow));
#endif
```

> What about keeping the minimize to tray feature only on Windows?

Sounds good to me.

**Comment by jonasschnelli** [comment]
Can reproduce the issue on Windows 8.1.
Tested this PR and it works,... however there is a tiny visual glich when minimising to tray (it show the main window for a couple of ms maximised after minimising), which is not really problematic.

OSX does not do minimise to tray... 
needs Linux testing.

**Comment by vajdaz** [comment]
I have made an additional change that hides the tray icon for non-Windows systems, disables the "Hide tray icon" and "Minimize to tray" options on the Settings' Winwow tab and sets the according settings in the OptionsModel (removing them from the configuration file).

On Windows, evertything works as before. On other systems, no tray icon is visible, and no tray icon settings are available. The "Minimize on close" option still exists on all OSes.

Should I create a new PR, or should I just add those changes to this one?

**Comment by laanwj** [comment]
Probably best to do so as a new PR, as the change is not directly pertinent to this one.

**Comment by laanwj** [comment]
Last reply from author is some time last year, closing and adding "Up for grabs" (let me know if you start work againand I should reopen).

**Pull request closed without merging**